src/encoded/static/components/visualizations.js

```diff
@@ -244,13 +246,273 @@ export class TestViz extends React.Component {
     }
 
     render() {
-
-        // const context = this.props.context;
-
         return (
             <div>
                 <div ref={(div) => { this.chartdisplay = div; }} className="chart-display" />
             </div>
         );
     }
 }
+
+BarChart.propTypes = {
+    context: PropTypes.object.isRequired,
+    dataFilter: PropTypes.string.isRequired,
+    chartWidth: PropTypes.number.isRequired,
+    chartLimit: PropTypes.number.isRequired, // limit is set to 0 for all the data
+    chartOrientation: PropTypes.string.isRequired,
+};
+
+function filterForBiosample(element, key) {
+    if (element.biosample_ontology) {
+        if (element.biosample_ontology.term_name) {
+            return element.biosample_ontology.term_name === key;
+        }
+        return false;
+    }
+    return false;
+}
+
+// Sanitize user input and facet terms for comparison: convert to lowercase, remove white space and asterisks (which cause regular expression error)
+const sanitizedString = inputString => inputString.toLowerCase()
+    .replace(/ /g, '') // remove spaces (to allow multiple word searches)
+    .replace(/[*?()+[\]\\/]/g, ''); // remove certain special characters (these cause console errors)
+
+// Display information on page as JSON formatted data
+export class ChartTable extends React.Component {
+    constructor(props, context) {
+        super(props, context);
+
+        this.state = {
+            chartData: [],
+            chartTitle: '',
+            chartMax: 0,
+            leftMargin: 0,
+            currentTarget: [],
+            data: [],
+            unsanitizedSearchTerm: '',
+            searchTerm: '',
+        };
+
+        // Bind `this` to non-React methods.
+        this.drawCharts = this.drawCharts.bind(this);
+        this.handleClick = this.handleClick.bind(this);
+        this.handleSearch = this.handleSearch.bind(this);
+        this.clearSearch = this.clearSearch.bind(this);
+        this.expandTerms = this.expandTerms.bind(this);
+        this.collapseTerms = this.collapseTerms.bind(this);
+    }
+
+    componentDidMount() {
+        this.drawCharts();
+    }
+
+    handleClick(clickID) {
+        if (!(this.state.currentTarget.includes(`table${clickID}`))) {
+            this.setState({
+                currentTarget: [...this.state.currentTarget, `table${clickID}`],
+            });
+        } else {
+            const removedTarget = this.state.currentTarget.filter(element => element !== `table${clickID}`);
+            this.setState({
+                currentTarget: removedTarget,
+            });
```

Comment:
The cart code does something like this to remove an element. I don’t think it’s guaranteed to perform faster than this, but filter does have to go through the entire array, and I don’t think this does. If currentTarget can’t get very long, then this probably doesn’t matter:

```
            const doomedIndex = state.elements.indexOf(action.elementAtId);
            if (doomedIndex !== -1) {
                return Object.assign({}, state, {
                    elements: state.elements
                        .slice(0, doomedIndex)
                        .concat(state.elements.slice(doomedIndex + 1)),
                });
            }
```

Reply:
It shouldn't get too long but I think slice() is a better choice anyway.

Reply:
`removedTarget` needs to be a memory copy of this.state.currentTarget to use with setState, where here you modify this.state (https://stackoverflow.com/questions/26253351/correct-modification-of-state-arrays-in-reactjs/#26254086). You’ll need to pass a complete copy of this.state.currentTarget with an element removed.